lint/long-chain_checker.go

```diff
@@ -0,0 +1,119 @@
+package lint
+
+import (
+	"fmt"
+	"go/ast"
+	"strings"
+
+	"github.com/Quasilyte/astcmp"
+)
+
+type longChainChecker struct {
+	ctx *Context
+}
+
+func newLongChainChecker(ctx *Context) Checker {
+	return &longChainChecker{
+		ctx: ctx,
+	}
+}
+
+func (c *longChainChecker) Check(f *ast.File) {
+	for _, decl := range f.Decls {
+		if decl, ok := decl.(*ast.FuncDecl); ok {
+			if decl.Body == nil {
+				continue
+			}
+			for _, stmt := range decl.Body.List {
+				if stmt, ok := stmt.(*ast.SwitchStmt); ok {
+					c.checkSwitch(stmt)
+				}
+			}
+		}
+	}
+}
+
+func (c *longChainChecker) checkSwitch(stmt *ast.SwitchStmt) {
+	exprs := []ast.Expr{}
+	for _, s := range stmt.Body.List {
+		cas := s.(*ast.CaseClause)
+		if cas.List == nil {
+			continue
+		}
+		for _, expr := range cas.List {
+			exprs = append(exprs, expr)
+		}
+	}
+	if len(exprs) < 2 {
+		return
+	}
+
+	cp := c.exprToList(exprs[0])
+	for _, e := range exprs[1:] {
+		cp = c.commonPrefix(cp, c.exprToList(e))
+	}
+
+	const n = 3
+
+	if len(cp) > n {
+		c.warn(cp, stmt)
+	}
+}
+
+func (c *longChainChecker) exprToList(expr ast.Expr) []ast.Expr {
+	res := []ast.Expr{}
+	tmp := expr
+	for {
+		switch t := tmp.(type) {
+		case *ast.SelectorExpr:
+			res = append(res, t.Sel)
+			tmp = t.X
+		default:
+			res = append(res, tmp)
+
+			for i := len(res)/2 - 1; i >= 0; i-- {
+				opp := len(res) - 1 - i
+				res[i], res[opp] = res[opp], res[i]
+			}
```

Comment:
I think it's copy/pasted slice reverse code.
I don't like `opp` name at all. Can it be `j`?
In addition, I would like it to be moved to `reverseExprSlice` method just to make intention more clear.